src/SimplePDPServiceWithPayments.sol

```diff
@@ -0,0 +1,778 @@
+        uint256 railId; // ID of the payment rail
+        address payer; // Address paying for storage
+        address payee; // SP's beneficiary address
+        uint256 commissionBps; // Commission rate for this proof set
+        string metadata; // General metadata for the proof set
+        string[] rootMetadata; // Array of metadata for each root
+        mapping(uint256 => string) rootIdToMetadata; // Mapping from root ID to its metadata
+    }
+
+    // Decode structure for proof set creation extra data
+    struct ProofSetCreateData {
+        string metadata;
+        address payer;
+    }
+
+    // Mappings
+    mapping(uint256 => uint256) public provingDeadlines;
+    mapping(uint256 => bool) public provenThisPeriod;
+    mapping(uint256 => ProofSetInfo) public proofSetInfo;
+
+    // Mapping from rail ID to proof set ID for arbitration
+    mapping(uint256 => uint256) public railToProofSet;
+
+    // Event for arbitration
+    event PaymentArbitrated(
+        uint256 railId,
+        uint256 proofSetId,
+        uint256 originalAmount,
+        uint256 modifiedAmount,
+        uint256 faultedEpochs
+    );
+
+    // Track which proving periods have valid proofs (proofSetId => periodId => isProven)
+    mapping(uint256 => mapping(uint256 => bool)) public provenPeriods;
+
+    // Track when proving was first activated for each proof set
+    mapping(uint256 => uint256) public provingActivationEpoch;
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(
+        address _pdpVerifierAddress,
+        address _paymentsContractAddress,
+        address _usdFcTokenAddress,
+        uint256 _initialOperatorCommissionBps
+    ) public initializer {
+        __Ownable_init(msg.sender);
+        __UUPSUpgradeable_init();
+
+        require(
+            _pdpVerifierAddress != address(0),
+            "PDP verifier address cannot be zero"
+        );
+        require(
+            _paymentsContractAddress != address(0),
+            "Payments contract address cannot be zero"
+        );
+        require(
+            _usdFcTokenAddress != address(0),
+            "USDFC token address cannot be zero"
+        );
+        require(
+            _initialOperatorCommissionBps <= COMMISSION_MAX_BPS,
+            "Commission exceeds maximum"
+        );
+
+        pdpVerifierAddress = _pdpVerifierAddress;
+        paymentsContractAddress = _paymentsContractAddress;
+        usdFcTokenAddress = _usdFcTokenAddress;
+        operatorCommissionBps = _initialOperatorCommissionBps;
+
+        // Read token decimals from the USDFC token contract
+        tokenDecimals = IERC20Metadata(_usdFcTokenAddress).decimals();
+
+        // Initialize the fee constants based on the actual token decimals
+        PROOFSET_CREATION_FEE = (1 * 10 ** tokenDecimals) / 10; // 0.1 USDFC
+        RATE_PER_MIB_PER_EPOCH = (1 * 10 ** tokenDecimals) / 10; // 0.1 USDFC per MiB per epoch
```

Comment:
This seems super expensive to me. If I am calculating correctly, the fee is 8,640 USDFC per MiB per month.

Assuming epoch = filecoin epoch (30 seconds):

```
 0.1 * 2 epochs/min * 60 mins/hour * 24 hours/day * 30 days/month = 8,640
```

Reply:
@bajtos This is just a "cookbook" contract to show to combine PDP + Payments. This contract will not be shipped to mainnet.